[Config] Document 'memtx' configuration settings (#4038)
tarantool/doc: doc/reference/configuration/configuration_reference.rst

    For example, :ref:`iproto <configuration_reference_iproto>` and :ref:`database <configuration_reference_database>` configuration parameters defined at the group level are applied to all instances in this group.

..  _configuration_reference_memtx:

> memtx

Reviewer: The entire `memtx` section should be after `instances`. Now it beaks the `groups -> replicasets -> instances hierarchy`. 

<img width="360" alt="image" src="https://github.com/tarantool/doc/assets/38073144/216a9411-9cc4-4cdb-8da1-385f354e4f2e">

Author: Thanks, fixed!

-----

> The ``memtx`` section is used to configure parameters related to :ref:`memtx engine <configuration_memtx>`.

Reviewer: ```suggestion
The ``memtx`` section is used to configure parameters related to the :ref:`memtx engine <configuration_memtx>`.
```

-   :ref:`memtx.slab_alloc_factor <configuration_reference_memtx_slab_alloc_factor>`
-   :ref:`memtx.slab_alloc_granularity <configuration_reference_memtx_slab_alloc_granularity>`
-   :ref:`memtx.sort_threads <configuration_reference_memtx_sort_threads>`

..  _configuration_reference_memtx_allocator:

..  confval:: memtx.allocator

    Specify the allocator used for memtx tuples.
    The possible values are ``system``  and ``small``:

    *   ``system`` is based on the ``malloc`` function.
        The allocator allocates memory as needed, checking that the quota is not exceeded.

    *   ``small`` is a special `slab allocator <https://github.com/tarantool/small>`_.
        The allocator repeatedly uses a memory block to allocate objects of the same type.
        Note that this allocator is prone to unresolvable fragmentation on specific workloads,
        so you can switch to ``system`` in such cases.

    |
    | Type: string
    | Default: 'small'
    | Environment variable: TT_MEMTX_ALLOCATOR

..  _configuration_reference_memtx_max_size:

..  confval:: memtx.max_tuple_size

    Size of the largest allocation unit for the memtx storage engine in bytes.

>     It can be increased if it is necessary to store large tuples.
>     See also: ``vinyl.max_tuple_size``.

Reviewer: Let's drop links to vinyl options. Looks like they are not relevant for memtx users.

```suggestion
    It can be increased if it is necessary to store large tuples.
```

    |
    | Type: integer
    | Default: 1048576
    | Environment variable: TT_MEMTX_MAX_TUPLE_SIZE

..  _configuration_reference_memtx_memory:

..  confval:: memtx.memory

    The amount of memory in bytes that Tarantool allocates to actually store tuples.
    When the limit is reached, :ref:`INSERT <box_space-insert>` and
    :ref:`UPDATE <box_space-insert>` requests fail with the  :errcode:`ER_MEMORY_ISSUE` error.
    The server does not go beyond the ``memtx.memory`` limit to allocate tuples, but there is additional memory

>     used to store indexes and connection information.

Reviewer: Let's include an example with setting a memory size here. Looks like this option is the most important.

    |
    | Type: integer
    | Default: 268435456
    | Environment variable: TT_MEMTX_MEMORY

..  _configuration_reference_memtx_min_size:

..  confval:: memtx.min_tuple_size

    Size of the smallest allocation unit in bytes.

>     It can be decreased if most of the tuples are very small.
>     The value must be between 8 and 1048280 inclusive.
> 
>     |
>     | Type: integer
>     | Default: 16
>     | Environment variable: TT_MEMTX_MIN_TUPLE_SIZE

Reviewer: ```suggestion
    It can be decreased if most of the tuples are very small.

    |
    | Type: integer
    | Default: 16
    | Possible values: between 8 and 1048280 inclusive
    | Environment variable: TT_MEMTX_MIN_TUPLE_SIZE
```

..  _configuration_reference_memtx_slab_alloc_factor:

..  confval:: memtx.slab_alloc_factor

    The multiplier for computing the sizes of memory
    chunks that tuples are stored in. A lower value may result in less wasted
    memory depending on the total amount of memory available and the
    distribution of item sizes. Allowed values range from 1 to 2.

    See also: :ref:`memtx.slab_alloc_granularity <configuration_reference_memtx_slab_alloc_granularity>`

    |
    | Type: number
    | Default: 1.05
    | Environment variable: TT_MEMTX_SLAB_ALLOC_FACTOR

..  _configuration_reference_memtx_slab_alloc_granularity:

..  confval:: memtx.slab_alloc_granularity

    Specify the granularity in bytes of memory allocation in the :ref:`small allocator <cfg_storage-memtx_allocator>`.
    The value should be a power of two and should be greater than or equal to 4.

>     Below are few recommendations on how to adjust the ``slab.alloc.granularity`` value:

Reviewer: ```suggestion
    Below are few recommendations on how to adjust the ``slab_alloc_granularity`` value:
```

    * To store small tuples of approximately the same size, set the option to 4 bytes to save memory.

>     * To store tuples of different sizes, you can increase the ``memtx.slab_alloc_granularity`` value.

Reviewer: ```suggestion
    * To store tuples of different sizes, you can increase the ``slab_alloc_granularity`` value.
```
